src/basic/random-util.c

```diff
@@ -68,8 +107,26 @@ int acquire_random_bytes(void *p, size_t n, bool high_quality_required) {
                          * a best-effort basis. */
                         have_syscall = true;
 
-                        if (!high_quality_required)
-                                return -ENODATA;
+                        if (!high_quality_required) {
+                                uint64_t u;
+                                size_t k;
+
+                                /* Try x86-64' RDRAND intrinsic if we have it. We only use it if high quality
+                                 * randomness is not required, as we don't trust it (who does?). Note that we only do a
+                                 * single iteration of RDRAND here, even though the Intel docs suggest calling this in
+                                 * a tight loop of 10 invocatins or so. That's because we don't really care about the
+                                 * quality here. */
+
+                                if (rdrand64(&u) < 0)
```

Comment:
Hmm, wouldn't it be better to just use `rdrand64` first if `!high_quality_required`, i.e. skip calling `getrandom` totally?

Reply:
so, i was tempted to do that, but so far our logic was that hashmaps may initially have crap entropy as long as it gets better later on, i.e. when they are under pressure. But if RDRAND can't be trusted, then it's neither good early *nor* late, hence i stepped away from it...

but dunno, maybe my paranoia on this is just paranoia and we should always use RDRAND if low-quality is requested, dunno? opinions?

Reply:
+1 on using RDRAND directly as a fallback.

+1 on trying to get entropy from `getrandom` first.

If the kernel is happy to give you randomness, then just take it from it.

There's also [Ted Ts'o's patch](https://lwn.net/Articles/760584/) that makes the kernel use more of RDRAND on startup, which I believe was just merged and might make it into many distros soon as well.

Cheers,
Filipe

Reply:
@filbranden using getrandom() first is not without drawbacks, as reading from it takes away entropy from the pool, and there usually is very very little entropy these days in the pool early on boot on many machines. going for RDRAND first would solve that of course, which i figure is what @keszybz is arguing for...